cachi2/core/package_managers/pip.py

```diff
@@ -1600,49 +1627,91 @@ def _validate_provided_hashes(requirements: list[PipRequirement], require_hashes
+        return {
+            "package": self.package,
+            "version": self.version,
+            "path": self.path,
+        }
 
-    :param PipRequirement requirement: PyPI requirement from a requirement.txt file
-    :param RootedPath pip_deps_dir: The deps/pip directory in a Cachi2 request bundle
-    :param str pypi_url: URL of the PyPI server or a proxy
-    :param (requests.auth.AuthBase | None) pypi_auth: Authorization for the PyPI server
 
-    :return: Dict with package name, version and download path
-    :raises FetchError: if PyPI query failed
-    :raises PackageRejected: if sdists for the package is not found or yanked
-    """
-    timeout = get_config().requests_timeout
+def _process_package_distributions(
+    requirement: PipRequirement,
+    pip_deps_dir: RootedPath,
+    user_checksums: dict[str, set[ChecksumInfo]] = dict(),
+    allow_binary: bool = False,
+) -> tuple[DistributionPackageInfo, list[DistributionPackageInfo]]:
     package = requirement.package
     version = requirement.version_specs[0][1]
+    canonical_version = canonicalize_version(version)
 
-    # See https://www.python.org/dev/peps/pep-0503/
-    package_url = f"{pypi_url.rstrip('/')}/simple/{canonicalize_name(package)}/"
+    client = PyPISimple()
     try:
-        pypi_resp = pkg_requests_session.get(package_url, auth=pypi_auth, timeout=timeout)
-        pypi_resp.raise_for_status()
-    except requests.RequestException as e:
+        timeout = get_config().requests_timeout
+        project_page = client.get_project_page(package, timeout)
+        packages = project_page.packages
+    except (requests.RequestException, NoSuchProjectError) as e:
         raise FetchError(f"PyPI query failed: {e}")
 
-    html = bs4.BeautifulSoup(pypi_resp.text, "html.parser")
-    # Find all anchors anywhere in the doc, the PEP does not specify where they should be
-    links = html.find_all("a")
+    wheels = []
+    if allow_binary:
+        filtered = [
+            p
+            for p in packages
+            if p.version
+            and canonicalize_version(p.version) == canonical_version
+            and p.package_type == "wheel"
+        ]
+
+        user_req_checksums = user_checksums.get(requirement.download_line, set())
+        for w in filtered:
+            dpi = DistributionPackageInfo(
+                package=package,
+                version=version,
+                path=pip_deps_dir.join_within_root(w.filename).path,
+                url=w.url,
+                is_yanked=w.is_yanked,
+            )
+
+            if len(w.digests) == 0:
+                dpi.checksums = user_req_checksums
+                wheels.append(dpi)
+                continue
+
+            for algorithm, digest in w.digests.items():
+                checksum = ChecksumInfo(algorithm, digest)
+                if checksum in user_req_checksums:
+                    dpi.checksums.add(checksum)
+                    wheels.append(dpi)
+                    break
```

Comment:
We need to make sure to inform the user when we filter out a wheel because of a checksum mismatch. An INFO-level log message would probably be appropriate

Reply:
what if none of the checksums matches ? Again, should we download all the wheels ?

Reply:
If the user specifies the expected checksums, we must not download (or at least must not leave in the output directory) anything that doesn't match

If nothing matches, we definitely shouldn't download anything and we should probably raise an error right away

Reply:
Not sure what should happen if only the sdist matches :thinking: 

Maybe just log a warning? If none of the wheels match and the sdist also doesn't match, then raise an error

Reply:
It's kind of a weird situation. If the user has `allow_binary: true`, the package has at least 1 wheel and none of the wheels match, that seems unintentional. But you never know what the user's intentions could be, pip's --hash system is too loose

Reply:
I did some refactoring, hope it helps...(I'm starting to get lost in it) :smile: